### ISSUE TITLE
Support ffmpeg 5.x/libavcodec 59

### ISSUE THREAD

**Issue by rien333** [pytorch/pytorch#72250]
### 🚀 The feature, motivation and pitch

[FFmpeg 5.0](https://ffmpeg.org/#pr5.0) released a few weeks ago, and with that came a new libavcodec. This, in turn, implies that a lot of caffe2's code fails to build (pytorch: v1.10.2). To give a quick sample of all the compilation errors: 

``` gcc-machine-description
FAILED: caffe2/CMakeFiles/torch_cpu.dir/video/video_decoder.cc.o
/usr/bin/ccache /usr/lib/ccache/bin/c++ -DADD_BREAKPAD_SIGNAL_HANDLER -DCPUINFO_SUPPORTED_PLATFORM=1...
/tmp/makepkg/python-pytorch/src/pytorch-1.10.2/caffe2/video/video_decoder.cc: In constructor ‘caffe2::VideoDecoder::VideoDecoder()’:
/tmp/makepkg/python-pytorch/src/pytorch-1.10.2/caffe2/video/video_decoder.cc:15:5: error: ‘av_register_all’ was not declared in this scope
   15 |     av_register_all();
      |     ^~~~~~~~~~~~~~~
/tmp/makepkg/python-pytorch/src/pytorch-1.10.2/caffe2/video/video_decoder.cc:16:5: error: ‘avcodec_register_all’ was not declared in this scope
   16 |     avcodec_register_all();
      |     ^~~~~~~~~~~~~~~~~~~~
/tmp/makepkg/python-pytorch/src/pytorch-1.10.2/caffe2/video/video_decoder.cc: In memberfunction ‘void caffe2::VideoDecoder::getAudioSample(AVPacket&, AVCodecContext*, AVFrame*, SwrContext*, caffe2::Callback&, const caffe2::Params&)’:
/tmp/makepkg/python-pytorch/src/pytorch-1.10.2/caffe2/video/video_decoder.cc:30:17: error: ‘avcodec_decode_audio4’ was not declared in this scope; did you mean ‘avcodec_decode_subtitle2’?
   30 |   auto result = avcodec_decode_audio4(
      |                 ^~~~~~~~~~~~~~~~~~~~~
      |                 avcodec_decode_subtitle2
/tmp/makepkg/python-pytorch/src/pytorch-1.10.2/caffe2/video/video_decoder.cc:193:21: error: ‘struct AVStream’ has no member named ‘codec’
  193 |             stream->codec->codec_type == AVMEDIA_TYPE_AUDIO &&
```

Rolling release distros like Arch Linux are close to releasing ffmpeg 5.0. 

I made this into a feature request because I'm not really sure how torch deals with non-backwards compatible ffmpeg releases.

### Alternatives

I think there's no clean solution to this for me. Either torch breaks, or, if I downgrade to ffmpeg 4, a lot of my other applications break (since a lot of stuff depends on ffmpeg/libavcodec)

### Additional context

Building the pytorch 1.10.2 release from source

**Comment by gspr** [comment]
I'm sorry, I don't know how I overlooked this issue when filing #72254. My apologies. I'll leave it up to the PyTorch team to decide which bug report should be closed in favor of the other.

For reference, Debian is also nearing its FFMPEG 5.0 transition, and [hitting the problem](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1004782).

**Comment by rien333** [comment]
Thanks for pointing out that Debian will also run into this problem. Goes to show that some kind of action needs to be undertaken.

**Comment by rien333** [comment]
Closed as a duplicate of #72254, which in some ways is more accurate.
